sieve_of_eratosthenes/SieveOfEratosthenes.java

```diff
@@ -1,34 +1,128 @@
-/*
- *  Following implementation of sieve of eratosthenes returns a boolean array which contains true is its index is prime
- *  less than or equal to a number n.
- *  Space complexity : O(n).
- *  Time complexity : O(n).
- */
+import java.util.Arrays;
 
+/**
+ * <p>This class represents a Sieve of Eratosthenes. It exposes an isPrime(int) method.</p>
+ *
+ * <p>When creating a Sieve, you need to specify the max number that you might query.</p>
+ *
+ * <p>Time complexity for creating the Sieve is O(n * n) but for querying is O(1). Space complexity is O(n)</p>
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes">Sieve of Eratosthenes</a>
+ */
 public class SieveOfEratosthenes {
 
-    public static boolean[] sieveOfEratosthenes(int n) {
-        int sqrtOfn = (int)Math.sqrt(n) + 1;
-        boolean[] primes = new boolean[n + 1];
-        for (int i = 2; i < n + 1; ++i) {
-            primes[i] = true;
-        }
-        for (int i = 2; i <= sqrtOfn; ++i) {
-            if (primes[i]) {
-                for (int j = 2 * i; j < n + 1; j += i) {
-                    primes[j] = false;
-                }
+    private boolean[] sieve;
+
+    /**
+     * Creates a sieve that can be queried for numbers up to maxNumber.
+     *
+     * @param maxNumber
+     */
+    public SieveOfEratosthenes(final int maxNumber) {
+        initSieve(maxNumber);
+        uncrossAllNumbers();
+        crossObviousNumbers();
+        crossMultiplesOfKnownPrimes();
```

Comment:
These three functions (lines 23, 24, 25) can be merged into one. This would make the algorithm look more self-explanatory as the array would be modified in a single function with all the logic.

Reply:
@singhpratyush when I think of not mixing levels of abstraction, is getting to code like this, that reads more like prose than like code. It makes it simple to understand for anyone that understand the domain of your problem. So I think is ok that to construct a sieve you understand there's a 4-step process. If I merge the methods, this will be lost. Also if you think in terms of SRP a method should only do 1 thing, I think having those 4 methods help them comply ti that principle.